examples/dcgan.py

```diff
@@ -162,12 +169,16 @@
 
 else:
     generator = None
+
 dataloader = torch.utils.data.DataLoader(
     dataset,
-    batch_size=opt.batch_size,
-    shuffle=True,
     num_workers=int(opt.workers),
     generator=generator,
+    batch_sampler=UniformWithReplacementSampler(
```

Comment:
Yeah, either with this sampler or with the default PyTorch sampler (see comment below), the sticking point is that currently we would leave configuring dataloading to the client, without Opacus helping you in any way.

This is fine and not something we will handle in this review, but we have been thinking for a while to have Opacus help on this too, either by attaching to a dataloader, or by accepting a dataloader in input and returning a new dataloader with all these settings fixed for you, etc.

We could really use extra input on this :)

Reply:
As explained previously, if we would like to evaluate the privacy budget with Mironov et al evaluation [1], we must use batches sampled according to the Sampled Gaussian Mechanism. As far as I know, this batch construction cannot be directly implemented with RandomSampler or other PyTorch samplers. Thus, I suggest to add this UniformWithReplacementSampler sampler in Opacus in order to make easier the use of the correct batch construction by the Opacus users.

[1] Rényi Differential Privacy of the Sampled Gaussian Mechanism

Reply:
Oh no, this comment was entirely about how to organize the API.

Whatever the sampler implementation used, currently the PrivacyEngine does not own your dataloader, which means that we leave configuring it properly entirely to the client. This was already a sore point before, because you would normally want to use a secure generator for it which is something that is easy to forget.